### PR TITLE
[Build/TensorRT] Add TensorRT dependency

### PR DESCRIPTION
**Pull request by dongju-chae**
This patch adds the build dependency for TensorRT.

CC: @again4you 

Signed-off-by: Dongju Chae <dongju.chae@samsung.com>

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -186,6 +186,28 @@ if not get_option('snpe-support').disabled()
   endif
 endif
 
+# tensorrt
+tensorrt_dep = dependency('', required: false)
```
Comment (wooksong) [inline meson.build:190]: Is this effective line?

Comment (zhoonit) [reply meson.build:190]: I think a discussion made at https://github.com/nnstreamer/nntrainer/pull/150#discussion_r439172394 might help this line.

File: meson.build [hunk 2]
```diff
@@ -245,6 +267,9 @@ features = {
   },
   'protobuf-support': {
     'extra_deps': [ pb_comp ]
+  },
+  'tensorrt-support': {
+    'extra_deps': tensorrt_dep
```
Comment (myungjoo) [inline meson.build:272]: It's ```extra_deps```, not ```extra_dep```; thus, ```[ tensorrt_dep ]``` will look more beautiful :)

File: meson.build [hunk 3]
```diff
@@ -186,6 +186,28 @@ if not get_option('snpe-support').disabled()
...
+  cuda_vers = [
+    '11.0',
+    '10.2',
+  ]
+
+  foreach ver : cuda_vers
+    cuda_dep = dependency('cuda-' + ver, required: false)
+    cudart_dep = dependency('cudart-' + ver, required: false)
+    if cuda_dep.found() and cudart_dep.found()
+      break
```
Comment (zhoonit) [inline meson.build:201]: is this okay if they are not found for all `cuda_vers` ?

Comment (zhoonit) [reply meson.build:201]: self answered: yes. it is handled at line 269

File: meson.build [hunk 4]
```diff
@@ -186,6 +186,28 @@ if not get_option('snpe-support').disabled()
   endif
 endif
 
+# tensorrt
+tensorrt_dep = dependency('', required: false)
+if not get_option('tensorrt-support').disabled()
+  cuda_vers = [
+    '11.0',
+    '10.2',
+  ]
```
Comment (myungjoo) [inline meson.build:195]: Do you need to restrict CUDA versions?

File: meson.build (hunk 2 above)
Comment (dongju-chae) [reply meson.build:272]: Ah, `tensorrt_dep` is actually 'deps'. I will rename it.

File: meson.build (hunk 1 above)
Comment (wooksong) [reply meson.build:190]: @zhoonit I could not still get it. 
```suggestion
tensorrt_dep = disabler()
```

Do you mean this?

File: meson.build [hunk 5]
```diff
@@ -186,6 +186,28 @@ if not get_option('snpe-support').disabled()
...
+    cudart_dep = dependency('cudart-' + ver, required: false)
+    if cuda_dep.found() and cudart_dep.found()
+      break
+    endif
+  endforeach
+
+  nvinfer_lib = cxx.find_library('nvinfer', required: false)
+  nvinfer_dep = declare_dependency(dependencies: nvinfer_lib)
+
+  tensorrt_dep = [ nvinfer_dep, cuda_dep, cudart_dep ]
```
Comment (wooksong) [inline meson.build:208]: wih
```
tensorrt_dep = disabler()
```
How about this?
```suggestion
  tensorrt_dep = declare_dependency(
      ependencies: [ nvinfer_dep, cuda_dep, cudart_dep ]
  )
```

Comment (dongju-chae) [reply meson.build:208]: Looks good!

File: meson.build (hunk 4 above)
Comment (dongju-chae) [reply meson.build:195]: I'm not sure.. but as described in https://github.com/NVIDIA/TensorRT, it seems that cuda-11.0 and cuda-10.2 are recommended.

File: meson.build (hunk 2 above)
Comment (dongju-chae) [reply meson.build]: That's right, updated.

File: meson.build (hunk 1 above)
Comment (zhoonit) [reply meson.build:190]: I am not sure if a disabler object acts like a dependency object as the reference states that the disabler object has only one method: `found()`. If it is fine to pass a disabler object to a `build_target()`, it is equivalent.

Comment (wooksong) [reply meson.build:190]: @zhoonit If tensorrt_dep.fount() is false, why do we provide this tensorrt_dep to build_target()?

Comment (zhoonit) [reply meson.build:190]: I don't know the exact use case only seeing this line. So I am not sure if `disabler()` can substitute `dependency('', required:false)`

What I know is `found()` is required in this file which makes this line effective.

Although I can think of a usecase for passing not_found to the build_target

```meson
A = dependency('foo', required: false)
B = executable('bar', dependencies: A, fallback: [......])
```
or

```meson
nnfw_dep = dependency('', required: false)
if not get_option('nnfw-runtime-support').disabled()
  nnfw_dep = dependency('nnfw', required: false)
  if not nnfw_dep.found()
    nnfw_dep = cc.find_library('nnfw-dev', required: get_option('nnfw-runtime-support'))
  endif
endif
```
and adding `nnfw_dep` to a `build_target()` will safely throw when needed

File: meson.build (hunk 4 above)
Comment (myungjoo) [reply meson.build:195]: 1. Such info can be added as a comment.
2. Need some warning messages if CUDA is installed but the version is not supported (or not recommended), then.

File: meson.build (hunk 1 above)
Comment (wooksong) [reply meson.build:190]: @zhoonit
I hardly think so.
https://mesonbuild.com/Disabler.html

https://git.sailfishos.org/mirror/rygel/blob/134eb29fa1994ca2e97617579a28de0fb69d39bf/meson.build

disabler() is introduced for reducing such fragmented if-else-endif.

Comment (dongju-chae) [reply meson.build]: I think `disabler()` looks good. BTW, I have modified this PR to use `tensorrt_deps` not `tensorrt_dep`.

File: meson.build (hunk 4 above)
Comment (dongju-chae) [reply meson.build:200]: Got it, updated.

File: meson.build (hunk 1 above)
Comment (zhoonit) [reply meson.build]: Yes I think `disabler()` can replace the given line :)

File: meson.build [hunk 6]
```diff
@@ -186,6 +186,36 @@ if not get_option('snpe-support').disabled()
...
+if not get_option('tensorrt-support').disabled()
+  # check available cuda versions (11.0 and 10.2 are recommended)
+  cuda_vers = [
+    '11.0',
+    '10.2',
+    '10.1',
+    '10.0',
+    '9.2',
+    '9.1',
+    '9.0'
```
Comment (kparichay) [inline meson.build:199]: - as far as I know, if the tensorRT was compiled with a different version of cuda and cudnn, even though a different version of cuda/cudnn is available, it won't work (this is from what I experienced an year ago).
if this happens currently as well, I am not sure if these cuda version versions will help unless there is the specific cuda version check.
- as long as version checks are being added, cudnn version might also be helpful

Comment (dongju-chae) [reply meson.build:199]: That's right. Any installed TensorRT expects the exact version of cuda/cudnn. I will change the PR.

Comment (dongju-chae) [reply meson.build:199]: Hmm... Unfortunately, it's really hard (impossible) to find the corresponding cuda version from libnvinfer. Do you have idea?

Comment (kparichay) [reply meson.build:199]: I cant think of any way either.

**Review by kparichay — APPROVED** [review]
LGTM

**Review by wooksong — APPROVED** [review]
LGTM

**Review by myungjoo — APPROVED** [review]

**Pull request merged**